File: dsa_with_python/tree/binarytree/tree_traversals.py
from node import Node
from tree_builder import build_tree


class Traversals:
    def pre_order(self, root: Node):
        if root is None:
            print(-1, end=" ")
            return
        print(root.data, end=" ")
        self.pre_order(root.left)
        self.pre_order(root.right)

    def in_order(self, root: Node):
        if root is None:
            print(-1, end=" ")
            return
        self.in_order(root.left)
        print(root.data, end=" ")
        self.in_order(root.right)

    def post_order(self, root: Node):
        if root is None:
            print(-1, end=" ")
            return
        self.post_order(root.left)
        self.post_order(root.right)
        print(root.data, end=" ")

    def level_order(self, root: Node):
        if root is None:
            return
        q = []
        q.append(root)
        q.append(None)
        while len(q) > 1:
            node = q.pop(0)
            if node is None:
                print()
                q.append(None)
                continue
            print(node.data, end=" ")
            if node.left is not None:
                q.append(node.left)
            if node.right is not None:
                q.append(node.right)

    def level_order_2(self, root: Node):
        if root is None:
            return
        q = []
        q.append(root)
        while len(q) > 0:
            flag = len(q)
            for i in range(flag):
                node = q.pop(0)
                print(node.data, end=" ")
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            print()


if __name__ == "__main__":
    ob = Traversals()
    nodes = [1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1]
    root = build_tree(nodes)
    ob.pre_order(root)
    print()
    ob.in_order(root)
    print()
    ob.post_order(root)
    print()
    ob.level_order(root)

File: DsaWithCpp/Tree/BinaryTree/TreeTraversals.cpp
#include <iostream>
#include <queue>
#include "Node.cpp"
#include "TreeBuilder.cpp"
using namespace std;

class Traversals
{
public:
    static void preOrder(Node *root)
    {
        if (root == NULL)
        {
            cout << -1 << " ";
            return;
        }
        cout << root->data << " ";
        preOrder(root->left);
        preOrder(root->right);
    }

    static void inOrder(Node *root)
    {
        if (root == NULL)
        {
            cout << -1 << " ";
            return;
        }
        inOrder(root->left);
        cout << root->data << " ";
        inOrder(root->right);
    }

    static void postOrder(Node *root)
    {
        if (root == NULL)
        {
            cout << -1 << " ";
            return;
        }
        postOrder(root->left);
        postOrder(root->right);
        cout << root->data << " ";
    }
    
    // LevelOrder -> Approach 1
    static void levelOrder(Node *root)
    {
        if (root == NULL)
            return;
        queue<Node *> q;
        q.push(root);
        q.push(NULL);
        while (q.size() > 1)
        {
            Node *currentNode = q.front();
            q.pop();
            if (currentNode == NULL)
            {
                cout << endl;
                q.push(NULL);
                continue;
            }
            cout << currentNode->data << " ";
            if (currentNode->left != NULL)
                q.push(currentNode->left);
            if (currentNode->right != NULL)
                q.push(currentNode->right);
        }
    }
    
    // LevelOrder -> Approach 2
    static void levelOrder2(Node *root)
    {
        if (root == NULL)
            return;
        queue<Node *> q;
        q.push(root);
        while (!q.empty())
        {
            int flag = q.size();
            for (int i = 0; i < flag; i++)
            {
                Node *currentNode = q.front();
                q.pop();
                cout << currentNode->data << " ";
                if (currentNode->left != NULL)
                    q.push(currentNode->left);
                if (currentNode->right != NULL)
                    q.push(currentNode->right);
            }
            cout << endl;
        }
    }
};

// int main()
// {
//     Traversals ob;
//     int nodes[] = {1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1};
//     Node *root = buildTree(nodes);
//     ob.preOrder(root);
//     cout << endl;
//     ob.inOrder(root);
//     cout << endl;
//     ob.postOrder(root);
//     cout << endl;
//     ob.levelOrder(root);
//     return 0;
// }